### PR TITLE
accurately detect default gateways

### PR DESCRIPTION
**Pull request by aojea**
Address several critical edge cases in default interface detection and introduces a robust, rootless networking test framework.

1. Point-to-Point Interfaces: Removed the `r.Gw == nil` check. Previously, this caused the agent to completely ignore active VPNs and tunnels (like Wireguard or tun/tap) because P2P links route directly out of the device without a Gateway IP.
2. Route Metrics: The kernel relies on metrics (Priority) to determine the active route in a multi-WAN setup. The old code ignored this, returning all interfaces. It now correctly isolates the lowest metric for IPv4 and IPv6 independently, while preserving ECMP support.
3. Multipath Link Lookup: Fixed a bug where multipath routes were queried using the parent route's link index (which is often 0) rather than the nexthop's link index.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines:
   https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution
   and developer guide
   https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are
   addressing, especially if this is a release targeted pull request. For
   reference on required PR/issue labels, read here:
   https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR:
   https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If the PR is unfinished, see how to mark it:
   https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests

/kind bug

```release-note
Address several critical edge cases in default interface detection
```

Fixes: https://github.com/kubernetes-sigs/dranet/issues/152

### CODE REVIEW

File: pkg/inventory/net.go [hunk 1]
```diff
@@ -28,19 +30,25 @@ import (
 )
 
 // getDefaultGwInterfaces returns a set of interface names that are configured
-// as default gateways in the main routing table. It identifies these by querying
-// the main routing table for routes with an unspecified destination (0.0.0.0/0
-// for IPv4 or ::/0 for IPv6).
+// as active default gateways in the main routing table, respecting route metrics.
```
Comment (dkennetzoracle) [inline pkg/inventory/net.go:33]: just a note / side-effect, any route matching Dst == 0.0.0.0/0 or ::/0 in the main table now classifies the link as an uplink, which is a new semantic and probably worth a comment?

File: internal/testutils/userns.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
+package testutils
```
Comment (dkennetzoracle) [inline internal/testutils/userns.go:17]: do you need a license header? I see them everywhere else.

Comment (aojea) [reply internal/testutils/userns.go:17]: yes, good catch

File: pkg/inventory/net.go (hunk 1 above)
Comment (aojea) [reply pkg/inventory/net.go:33]: is that heuristic not correct?

Comment (aojea) [reply pkg/inventory/net.go:33]: it is the same behavior as before, isn;t it? 
it just expands the logic to take into consideration the weights, but previous logic about the 0.0.0.0 and ::/0 remains the same

Comment (dkennetzoracle) [reply pkg/inventory/net.go:33]: It is the same logic, but the route no longer has to be a gateway. Any route with scope=link for example but with no gateway but still matches 0.0.0.0/0 or ::/0 will also match here.

File: pkg/inventory/net_test.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+			name: "Independent families (IPv4 and IPv6 on different interfaces)",
+			setupRoutes: func() error {
+				if err := netlink.RouteAdd(&netlink.Route{Family: netlink.FAMILY_V4, Dst: defaultIPv4, Gw: gwIPv4, LinkIndex: links["eth0"].Attrs().Index, Priority: 100, Table: unix.RT_TABLE_MAIN}); err != nil {
+					return err
+				}
+				return netlink.RouteAdd(&netlink.Route{Family: netlink.FAMILY_V6, Dst: defaultIPv6, Gw: gwIPv6, LinkIndex: links["eth2"].Attrs().Index, Priority: 100, Table: unix.RT_TABLE_MAIN})
+			},
+			expectedResult: sets.New[string]("eth0", "eth2"),
+		},
+		{
```
Comment (dkennetzoracle) [inline pkg/inventory/net_test.go:144]: I'd add a test for Mixed-family default on a single interface. Basically, when you merge sets for ipv4 and ipv6 this asserts that the merge behaves (dual stacking an interface).
```go
{
    name: "Same interface wins both families",
    setupRoutes: func() error {
        if err := netlink.RouteAdd(&netlink.Route{Family: netlink.FAMILY_V4, Dst: defaultIPv4, Gw: gwIPv4, LinkIndex: links["eth0"].Attrs().Index, Priority: 100, Table: unix.RT_TABLE_MAIN}); err != nil {
            return err
        }
        return netlink.RouteAdd(&netlink.Route{Family: netlink.FAMILY_V6, Dst: defaultIPv6, Gw: gwIPv6, LinkIndex: links["eth0"].Attrs().Index, Priority: 100, Table: unix.RT_TABLE_MAIN})
    },
    expectedResult: sets.New[string]("eth0"),
},
```

File: pkg/inventory/net_test.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+			name: "Point-to-Point Interface (No Gateway IP)",
+			setupRoutes: func() error {
+				// Intentionally leaving Gw == nil here because Point-to-Point links don't have gateways
+				if err := netlink.RouteAdd(&netlink.Route{Family: netlink.FAMILY_V4, Dst: defaultIPv4, LinkIndex: links["wg0"].Attrs().Index, Priority: 50, Scope: netlink.SCOPE_LINK, Table: unix.RT_TABLE_MAIN}); err != nil {
+					return err
+				}
+				return netlink.RouteAdd(&netlink.Route{Family: netlink.FAMILY_V4, Dst: defaultIPv4, Gw: gwIPv4, LinkIndex: links["eth0"].Attrs().Index, Priority: 100, Table: unix.RT_TABLE_MAIN})
+			},
+			expectedResult: sets.New[string]("wg0"),
+		},
```
Comment (dkennetzoracle) [inline pkg/inventory/net_test.go:181]: Would it be worth also having a pinned Priority: 0 for P2P case?

You re using Prio 50 for wg0 and this exercises the GW == nil removal as well as the the lowest metric wins logic. However, it doesn't exercise a specific case where a P2P route has Priority: 0 which is the ipv4 default for a route installed without an explicit metric. This case would just lock in that 0 is a real comparable value.

OK with not doing it, it just feels a bit more explicit and defensive.

```go
{
    name: "P2P default with Priority 0 wins over gateway'd default",
    setupRoutes: func() error {
        if err := netlink.RouteAdd(&netlink.Route{Family: netlink.FAMILY_V4, Dst: defaultIPv4, LinkIndex: links["wg0"].Attrs().Index, Priority: 0, Scope: netlink.SCOPE_LINK, Table: unix.RT_TABLE_MAIN}); err != nil {
            return err
        }
        return netlink.RouteAdd(&netlink.Route{Family: netlink.FAMILY_V4, Dst: defaultIPv4, Gw: gwIPv4, LinkIndex: links["eth0"].Attrs().Index, Priority: 100, Table: unix.RT_TABLE_MAIN})
    },
    expectedResult: sets.New[string]("wg0"),
},
```

File: internal/testutils/userns.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+var (
+	isSupported bool
+	checkOnce   sync.Once
+)
+
+// IsSupported checks if unprivileged user namespaces are enabled on the host system.
+// It caches the result after the first check.
+func IsSupported() bool {
+	checkOnce.Do(func() {
+		cmd := exec.Command("sleep", "1")
```
Comment (dkennetzoracle) [inline internal/testutils/userns.go:37]: A small nit: this requires `sleep` to be on the system in `PATH`. If it's not, you'd get an error and it would be because of a missing binary not because user namespaces are not supported. Could check it?

File: internal/testutils/userns.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	if !IsSupported() {
+		t.Skip("Unprivileged user namespaces are not supported on this system")
+	}
+
+	// 3. Prepare the command to re-execute the current test binary.
+	cmd := exec.Command(os.Args[0])
+	cmd.Args = []string{os.Args[0], "-test.run=" + t.Name() + "$", "-test.v=true"}
+
+	for _, arg := range os.Args {
+		if strings.HasPrefix(arg, "-test.testlogfile=") {
```
Comment (dkennetzoracle) [inline internal/testutils/userns.go:86]: would there be any reason to add `-test.timeout` here? Any of these long?

**Review by dkennetzoracle — APPROVED** [review]
Some nits / questions, but I think it's a great add and a correctness fix. I'm not 100% sure it fixes the `NonUplinkChecker` situation on #138 on OKE, I'd need to double check. RA injects IPv6 defaults for us. If they share a metric (which I think they do) I still need something like NonUplinkChecker. If they don't I can drop NonUplinkChecker.

This definitely addresses one of the sub-issues in that PR, and is correct. LGTM!

**Comment by gauravkghildiyal** [comment]
> I'm not 100% sure it fixes the NonUplinkChecker situation on https://github.com/kubernetes-sigs/dranet/pull/138 on OKE, I'd need to double check.

@dkennetzoracle I'm very much interested in hearing about this after you've had the opportunity to try this out

**Comment by dkennetzoracle** [comment]
@gauravkghildiyal - I will let you know! It's very hard for me to get access to GB200+ so I need to make the most of my time on them haha. I'm pretty sure each NIC gets a default route for ipv6 and these use the kernel's RA priority (so 1024) so we wouldn't be able to differentiate the 8 RDMA NICs from each other by metric. They all look like peer's from he kernel's pov.

So they'd all get excluded still here, haha. But it should work for the Azure case

File: pkg/inventory/net.go (hunk 1 above)
Comment (anson627) [reply pkg/inventory/net.go:33]: +1 on semantic broadening — "default route with a gateway" to "any default route"

File: internal/testutils/userns.go (hunk 5 above)
Comment (anson627) [reply internal/testutils/userns.go:37]: maybe `exec.Command(os.Args[0], "-test.run=^$")` to avoid the external sleep dependency

File: pkg/inventory/net.go (hunk 1 above)
Comment (aojea) [reply pkg/inventory/net.go:33]: what is the difference, from the routing perspective the fact of having a next hop direction or just send to via the corresponding interface is not important, what is important is that any packet in the host will go through that specific interface ... that is why we don't want to expose it in the ResourceSlice

Comment (dkennetzoracle) [reply pkg/inventory/net.go:33]: I don't feel super strongly about it, (to me) it just adds clarity.

**Comment by anson627** [comment]
I just verified this PR on AWS p4d.24large, the primary NiC with default gateway/route (e.g. ens32) is excluded, and 4 EFA/ENA devices are properly included in the resource slice:

  | Device           | Type | Interface | PCI Address  | NUMA | IP                  |
|------------------|------|-----------|--------------|------|---------------------|
| pci-0000-10-01-0 | ENA  | ens33     | 0000:10:01.0 | 0    | 192.168.157.168/19  |
| pci-0000-10-1b-0 | EFA  | —         | 0000:10:1b.0 | 0    | rdmap16s27          |
| pci-0000-20-01-0 | ENA  | ens65     | 0000:20:01.0 | 0    | 192.168.155.210/19  |
| pci-0000-20-1b-0 | EFA  | —         | 0000:20:1b.0 | 0    | rdmap32s27          |
| pci-0000-90-01-0 | ENA  | ens129    | 0000:90:01.0 | 1    | 192.168.145.129/19  |
| pci-0000-90-1b-0 | EFA  | —         | 0000:90:1b.0 | 1    | rdmap144s27         |
| pci-0000-a0-01-0 | ENA  | ens161    | 0000:a0:01.0 | 1    | 192.168.132.181/19  |
| pci-0000-a0-1b-0 | EFA  | —         | 0000:a0:1b.0 | 1    | rdmap160s27         |

**Comment by anson627** [comment]
verified this PR on Azure GB300, the primary NiC (eth0) with default gateway is excluded, while Azure accelerated networking VF (e.g. enP8051s1) together with other mlx5_* NiCs are included

| # | Name             | RDMA Device | PCI Address  | NUMA | Vendor   | RDMA |
|---|------------------|-------------|--------------|------|----------|------|
| 1 | pci-0101-00-00-0 | mlx5_0      | 0101:00:00.0 | 0    | Mellanox | true |
| 2 | pci-0102-00-00-0 | mlx5_1      | 0102:00:00.0 | 0    | Mellanox | true |
| 3 | pci-0103-00-00-0 | mlx5_2      | 0103:00:00.0 | 1    | Mellanox | true |
| 4 | pci-0104-00-00-0 | mlx5_3      | 0104:00:00.0 | 1    | Mellanox | true |
| 5 | pci-1f73-00-02-0 | —           | 1f73:00:02.0 | 0    | Mellanox | true |

/lgtm

**Review by anson627 — APPROVED** [review]

**Comment by gauravkghildiyal** [comment]
Holding to have time to go through this and allow resolution of a few comments.

/hold

**Comment by aojea** [comment]
> RA injects IPv6 defaults for us. If they share a metric (which I think they do) I still need something like NonUplinkChecker. If they don't I can drop NonUplinkChecker.

From a kubernetes architectural and security perspective, we like to have certain guardrails so users can not shoot themselves on the feet. Any logic that allows an interface with a default route to be unmounted or modified based on an opportunistic RA injection is very risky.:

- In the Kubernetes networking model, there is typically no difference between the control plane and the dataplane traffic. If an interface intended for the dataplane receives a default route via RA, it will be 'hijacking' control plane traffic, so the kubelet and pods trying to connect to the apiserver will fail (unless this secondary interface has access too, but this looks a very convoluted setup with routing loops). This will effectively disconnect the entire node from the cluster.

- We already were hit by security issues "abusing" RA, so right now is RA processing disabled by default in all container interfaces, the same attack can be performed at the node level

> A vulnerability was found in all versions of containernetworking/plugins before version 0.8.6, that allows malicious containers in Kubernetes clusters to perform man-in-the-middle (MitM) attacks. A malicious container can exploit this flaw by sending rogue IPv6 router advertisements to the host or other containers, to redirect traffic to the malicious container. ([CVE-2020-10749](https://www.cve.org/CVERecord?id=CVE-2020-10749))

I still think we need a better mechanism for improving filtering, so I encourage you to check @gauravkghildiyal proposal in https://github.com/kubernetes-sigs/dranet/issues/152 , that I think will be able to accomodate the `NonUplinkChecker` functionality perfectly

**Comment by dkennetzoracle** [comment]
@aojea aligned with @gauravkghildiyal 's issue, it seems like a good fix and a generally better way to approach the problem!

File: pkg/inventory/net.go (hunk 1 above)
Comment (aojea) [reply pkg/inventory/net.go:33]: fair

File: pkg/inventory/net_test.go [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+					Gw:        gwIPv4, // Added Gateway
+					LinkIndex: links["eth0"].Attrs().Index,
+					Priority:  100,
+					Table:     unix.RT_TABLE_MAIN,
+				})
+			},
+			expectedResult: sets.New[string]("eth0"),
+		},
+		{
+			name: "Lower metric wins (IPv4)",
```
Comment (tamilmani1989) [inline pkg/inventory/net_test.go:105]: should we add one test for ipv6 with 2 default routes?

**Review by tamilmani1989 — APPROVED** [review]
changes lgtm. left a comment, i will leave to your discretion

File: internal/testutils/userns.go (hunk 6 above)
Comment (aojea) [reply internal/testutils/userns.go:86]: this setup was been working for more than one year in other projects like kube-network-policies, so I always prefer to wait for evidence to do changes rather than premature optimizations

File: internal/testutils/userns.go (hunk 5 above)
Comment (aojea) [reply internal/testutils/userns.go:37]: same as https://github.com/kubernetes-sigs/dranet/pull/153#discussion_r3102972405

**Comment by aojea** [comment]
pushed addressing latest comments, great reviews, caught some flaws:
- clarify function scope explicitly indicating it queries for the unspecified routes 0.0.0.0/0, ::/0 or nil that are the kernel default networks
- adding new unit tests

File: pkg/inventory/net_test.go (hunk 7 above)
Comment (aojea) [reply pkg/inventory/net_test.go:105]: added

File: pkg/inventory/net_test.go (hunk 3 above)
Comment (aojea) [reply pkg/inventory/net_test.go:144]: added

File: pkg/inventory/net_test.go (hunk 4 above)
Comment (aojea) [reply pkg/inventory/net_test.go:181]: added as "P2P default with Priority 0 wins over gateway default"

**Comment by dkennetzoracle** [comment]
@aojea @gauravkghildiyal I can confirm this fixed our issues (besides NIC orphaning) on OCI. I pulled in our OKE changes to this branch locally and deployed, and didn't need any additional changes. We made a few additional changes on our side, and fixed our dual ipv6 stack deployment. With these changes, the only thing we are missing is the cri upstream changes.

**Comment by aojea** [comment]
/hold cancel

based on the feedback from OKE and Azure, @gauravkghildiyal for final review

**Comment by gauravkghildiyal** [comment]
Perfect, thanks for validating @anson627 and @dkennetzoracle 

Thanks for the change @aojea 

/lgtm
/approve
